### ISSUE TITLE
bazel 0.8.0 generates different set of dynamically linked library from previous version

### ISSUE THREAD

**Issue by fbriggs** [bazelbuild/bazel#4211]
* Operating System: OS X 10.12.6
* Bazel version
0.5.4 vs 0.8.0 (or 0.7.0 vs 0.8.0)
* bazel info release
==>
release 0.8.0-homebrew

# Summary

I have a project where bazel is being used to compile an executable on OS X, which depends on several libraries include glfw, opencv, and others. For the sake of confidentiality, I have renamed that program 'some_program_compiled_with_bazel'. With previous versions of bazel, this was working fine, but as of 0.8.0, attempting to run the program gives the following runtime linker error:

Reason: Incompatible library version: some_program_compiled_with_bazel requires version 3.3.0 or later, but libglfw.3.2.dylib provides version 3.0.0

This issue does not occur when compiling the exact same code, BUILD, WORKSPACE, etc files on bazel 0.7.0 or bazel 0.5.4.

Running otool -L to inspect the libraries that the program is trying to link shows very different results between the executable compiled on bazel 0.5.4 vs. 0.8.0; the output looks suspicious for 0.8.0. Outputs from otool are listed below:

# compile with bazel 0.5.4

otool -L ./bazel-bin/some_program_compiled_with_bazel

	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_aruco.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_calib3d.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_ccalib.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_core.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_features2d.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_flann.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_fuzzy.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_highgui.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_imgcodecs.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_imgproc.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_ml.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_objdetect.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_optflow.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_photo.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_plot.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_reg.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_shape.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_stereo.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_stitching.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_superres.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_surface_matching.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_tracking.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_video.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_videoio.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_videostab.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_xfeatures2d.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_ximgproc.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/opencv/lib/libopencv_xphoto.3.3.0.dylib (compatibility version 3.3.0, current version 3.3.0)
	/private/var/tmp/_bazel_fbriggs/a2f69eca3eb9035c203eb7c1d55eaf4a/external/glfw/lib/libglfw.3.2.dylib (compatibility version 3.0.0, current version 3.2.0)
	/usr/local/opt/suite-sparse/lib/libcholmod.3.0.11.dylib (compatibility version 3.0.0, current version 3.0.11)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 307.5.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)

# compile with bazel 0.8.0

otool -L ./bazel-bin/some_program_compiled_with_bazel

	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.3.0, current version 3.3.0)
	@loader_path/../_solib_darwin/_U@glfw_S_S_Cglfw___Uexternal_Sglfw_Slib/libglfw.3.2.dylib (compatibility version 3.0.0, current version 3.2.0)
	/usr/local/opt/suite-sparse/lib/libcholmod.3.0.11.dylib (compatibility version 3.0.0, current version 3.0.11)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 307.5.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)

**Comment by iirina** [comment]
How do you get the `some_program_compiled_with_bazel` binary? Is it a `cc_binary`? How do you specify the `glfw` dependency?

**Comment by fbriggs** [comment]
@iirina - yes, it is a cc_binary

here is the contents of our glfw.BUILD file:

config_setting(
  name = "linux",
  values = { "cpu": "k8" },
)
config_setting(
  name = "osx",
  values = { "cpu": "darwin" },
)

```
cc_library(
  name = "glfw",
  srcs = glob(["lib/*.a", "lib/*.so*", "lib/*.dylib*"]),
  hdrs = glob(["include/GLFW/*"]),
  includes = ["include"],
  visibility = ["//visibility:public"],
  linkopts = select({":linux": ["-ldl", "-lGL", "-lGLU", "-lX11", "-lXrandr", "-lXxf86vm", "-lXinerama", "-lXcursor", "-lpthread"],
                     "//conditions:default": []}),
  linkstatic = 0,
)
```

**Comment by iirina** [comment]
@fbriggs thanks!
cc: @lberki do you have any thoughts on this?

**Comment by fbriggs** [comment]
I was hoping to work around this by statically linking glfw, but the discussion in (https://github.com/bazelbuild/bazel/issues/2055) appears to suggest that static linking doesnt work on os x.. ugh. so basically bazel 0.8.0 cannot link glfw statically or dynamically

**Comment by fbriggs** [comment]
Here's a minimal test case where I only compile glfw (and none of the rest of our project). In this case, it is working fine on OS X and with bazel 0.8.0.
[glfw_minimal.zip](https://github.com/bazelbuild/bazel/files/1528929/glfw_minimal.zip)

So this seems to suggest the problem I am seeing is more complicated / quite possible I have not provided enough info to see what's going on in the notes above.

**Comment by fbriggs** [comment]
after digging into this more, I have found that it doesn't reproduce in bazel 0.8.0 in a simplified test case consisting of an executable which depends only on GLFW. instead, it seems to be related to the list of dependencies for the cc_binary. In particular, the order of the list of deps for the cc_binary matters! if I move glfw to be first in the list of deps, then inspect the resulting executable with otool, the list of linker dependencies is very different.

question for the bazel devs: should the order of deps for a cc_binary ever matter?

**Comment by mhlopko** [comment]
Yes. C++ linking is (extremely :) order-dependent, so bazel retains the order of libs like you provided in the BUILD file. This way you can reshuffle libs so the binary links correctly. In the ideal case the order is expressed by the dependency hierarchy - if something A needs symbols from something else B, there should be a dependency edge in the graph from A to B. Can you isolate this case in your graph to see what's going on?

**Comment by mhlopko** [comment]
Any update? Can this be closed?

**Comment by fbriggs** [comment]
updates:
1) I re-tested on bazel 0.8.1 and the issue is still present
2) I managed to isolate the problem down to just an interaction between OpenCV and GLFW. I might be able to send out a minimal reproducing test case without sharing any confidential code
3) I ran bazel build -s and compared the results between bazel 0.5.4 and 0.8.1, and there are some minor difference in the link step, which I don't fully understand. the main differences that I can see is that things go do local-fastbuild in bazel 0.5.4 and darwin-fastbuild in bazel 0.8.1

please don't close just yet. ill see if I can put up a minimal reproducing test case today

**Comment by mhlopko** [comment]
Ha and we have a winner :) So the thing that changed is the introduction of ObjC crosstool. If you have xcode installed, bazel will pick a crosstool that supports both C++ and ObjC. If you want to revert back to the C++-only crosstool, subscribe to https://github.com/bazelbuild/bazel/issues/4231.

**Comment by fbriggs** [comment]
uuugh.. earlier today I had things working, them installed Xcode 9 and now that has broken everything. will still try to provide reproducing test case if I can, but now that's one more step away

**Comment by mhlopko** [comment]
You might want to accept xcode licence, and run it at least once to install some extensions. Then run bazel clean --expunge and you should be good to go.

**Comment by fbriggs** [comment]
I tried accepting the license and running bazel clean --expunge. that reduced the number of errors, but now I am still getting:

clang: error: cannot specify -o when generating multiple output files

**Comment by mhlopko** [comment]
Hmm can you share the output of bazel build //your-target --subcommands?

**Comment by fbriggs** [comment]
Here's a minimal test case:
[test.zip](https://github.com/bazelbuild/bazel/files/1549371/test.zip)

Before I installed Xcode 9, this was actually working with bazel 0.8.1, although something similar wasn't working in my larger repo (the difference I think was how opencv was brought in- through brew, or compiled with CMake). Now after installing Xcode 9, even this simplified test case doesn't work.

anyway, here's the answer to your request above:


bazel_glfw_opencv fbriggs$ bazel build //:all --subcommands
INFO: Analysed target //:TestGLFW (0 packages loaded).
INFO: Found 1 target...
SUBCOMMAND: # //:TestGLFW [action 'Compiling TestGLFW.cpp']
(cd /private/var/tmp/_bazel_fbriggs/114e6ac6240c9fab002bf2cadb1c5203/execroot/__main__ && \
  exec env - \
    APPLE_SDK_PLATFORM=MacOSX \
    APPLE_SDK_VERSION_OVERRIDE=10.13 \
    PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/fbriggs/Library/Python/3.6/bin:/usr/local/opt/clang-format@3.8/bin \
    TMPDIR=/var/folders/lt/8tbxs37s2kq6b8q07_2bpfs80000gp/T/ \
    XCODE_VERSION_OVERRIDE=9.2.0 \
  external/local_config_cc/wrapped_clang '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -fno-omit-frame-pointer -O0 -DDEBUG '-std=c++11' -iquote . -iquote bazel-out/darwin-fastbuild/genfiles -iquote external/glfw -iquote bazel-out/darwin-fastbuild/genfiles/external/glfw -iquote external/bazel_tools -iquote bazel-out/darwin-fastbuild/genfiles/external/bazel_tools -iquote external/opencv -iquote bazel-out/darwin-fastbuild/genfiles/external/opencv -isystem external/glfw/include -isystem bazel-out/darwin-fastbuild/genfiles/external/glfw/include -isystem 'external/glfw/include/GLFW/*' -isystem 'bazel-out/darwin-fastbuild/genfiles/external/glfw/include/GLFW/*' -isystem external/bazel_tools/tools/cpp/gcc3 -isystem external/opencv/include -isystem bazel-out/darwin-fastbuild/genfiles/external/opencv/include -isystem 'external/opencv/include/core/*' -isystem 'bazel-out/darwin-fastbuild/genfiles/external/opencv/include/core/*' -MD -MF bazel-out/darwin-fastbuild/bin/_objs/TestGLFW/TestGLFW.d '-frandom-seed=bazel-out/darwin-fastbuild/bin/_objs/TestGLFW/TestGLFW.o' '-isysroot __BAZEL_XCODE_SDKROOT__' -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c TestGLFW.cpp -o bazel-out/darwin-fastbuild/bin/_objs/TestGLFW/TestGLFW.o)
ERROR: /Users/fbriggs/Desktop/bazel_glfw_opencv/BUILD:1:1: C++ compilation of rule '//:TestGLFW' failed (Exit 1)
clang: error: cannot specify -o when generating multiple output files
Target //:TestGLFW failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.214s, Critical Path: 0.11s
FAILED: Build did NOT complete successfully

**Comment by fbriggs** [comment]
completely removing xcode got me back to a working system (instructions here: https://stackoverflow.com/questions/31011062/how-to-completely-uninstall-xcode-and-clear-all-settings)

**Comment by fbriggs** [comment]
i think we can close this issue. i can't repro it in my simplified test case under bazel 0.8.1, after uninstalling xcode 9.

there is still a difference in behavior between bazel 0.7.0 and 0.8.0 which is breaking the build in my repo, but I can't share enough of it to give a reproducing test case without exposing confidential code

**Comment by mhlopko** [comment]
Ok, uploaded a fix for #4231 internally, should be live soon. Closing this issue for now. Thanks for reporting.
